app/search_builders/collection_search_builder.rb

```diff
@@ -9,9 +11,9 @@ def include_item_ids(solr_parameters)
     solr_parameters[:fq] << "hasCollectionMember_ssim:#{item.id}"
   end
 
-  def initialize(item, processor_chain, scope)
-    super(processor_chain, scope)
-    @item = item
+  def include_contained_files(solr_parameters)
+    solr_parameters[:fq] ||= []
+    solr_parameters[:fq] << "{!join from=generic_files_tesim to=id}{!join from=hasCollectionMember_ssim to=id}id:#{collection.id}"
```

Comment:
In general, Solr joins are not entirely like SQL joins.  They have stronger limitations: 
- only one "table" (solr core), so more like a self-join
- no fields returned out of the "from" docs

But JOINs are typically a place Solr performance goes to die.  Given this community's tendency to treat Solr as a far-away thing to be abstracted, encapsulated and otherwise misunderstood, I would regard the introduction of joins with skepticism.  Basically, it ain't a RDBS.  

Clues from Lucid: https://lucidworks.com/blog/solr-and-joins/

If I understand correctly that the current modeling has object-documents and collection-documents (read "document" here as "Solr document") in the same core, i.e. that we are joining on UNIQUE_ID, then we should expect worst case performance similar to the post's example.  My interpretation of that is that it would be acceptable on the scale of say, collections (~100's of IDs), but unacceptable on the scale of every-object-in-repo (~1M).